main/src/library/LazyList.flix

```diff
@@ -181,4 +169,101 @@ namespace LazyList {
         case LCons(x, xs) => mapEAcc(f, force xs, ECons(f(x), acc))
         case LList(xs)    => mapEAcc(f, force xs, acc)
     }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `l` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, x1), x2)..., xn)`.
+    ///
+    pub def foldLeft(f: (b, a) -> b & ef, s: b, l: LazyList[a]): b & ef = match l {
+        case ENil         => s
+        case ECons(x, xs) => foldLeft(f, f(s, x),       xs)
+        case LCons(x, xs) => foldLeft(f, f(s, x), force xs)
+        case LList(xs)    => foldLeft(f, s,       force xs)
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `l` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(x1, ...f(xn-1, f(xn, s))...)`.
+    ///
+    pub def foldRight(f: (a, b) -> b & ef, s: b, l: LazyList[a]): b & ef = foldRightAcc(f, s, l, s1 -> s1 as & ef)
+
+    ///
+    /// Helper function for `foldRight`.
+    ///
+    def foldRightAcc(f: (a, b) -> b & ef, s: b, l: LazyList[a], g: b -> b & ef): b & ef = match l {
+        case ENil         => g(s)
+        case ECons(x, xs) => foldRightAcc(f, s,       xs, s1 -> g(f(x, s1)))
+        case LCons(x, xs) => foldRightAcc(f, s, force xs, s1 -> g(f(x, s1)))
+        case LList(xs)    => foldRightAcc(f, s, force xs, g)
+    }
+
+    ///
+    /// Returns `l` as an `Array`.
+    ///
+    /// Forces the entire list `l`.
+    ///
+    pub def toArray(l: LazyList[a]): Array[a] & Impure = match head(l) {
+        case None    => []
+        case Some(x) =>
+            let a = Array.new(x, length(l));
+            let f = (i, y) -> { a[i] = y; i + 1 };
+            foldLeft(f, 0, l);
+            a
+    }
+
+    ///
+    /// Returns `l` as a `List`.
+    ///
+    /// Forces the entire list `l`.
+    ///
+    pub def toList(l: LazyList[a]): List[a] = toListAcc(l, Nil) |> List.reverse
```

Comment:
Could all these conversion functions just be done with `fold`s?

Reply:
I have somehow convinced myself that it would be slightly faster to skip the call to fold and implement the iteration directly thus avoiding a couple of function calls. Is this a false assumption? If so, I'd agree with your comment.

Reply:
Yes, I think you are right that there may be a performance difference, though I'm not sure how significant it would be.